src/mlpack/methods/hdbscan/hdbscan_impl.hpp

```diff
@@ -0,0 +1,763 @@
+  arma::Mat<size_t> result;
+  GetClusters(condensedTree, result);
+
+  // save results obtained
+  assignments = result;
+}
+
+//  A class to help merge two clusters to form a
+// single linkage tree.
+class UnionForSingleLinkageClustering
+{
+  arma::Col<size_t> parentVector;
+  arma::Col<size_t> sizeVector;
+  size_t nextLabel;
+
+ public:
+  UnionForSingleLinkageClustering(size_t n)
+  {
+    // Initialize the elments
+    // There would be a max of 2*n-1 clusters
+    // formed after completion
+    // Initally all n points form a cluster
+    // After first merging there are
+    // (n-2) old clusters and 1 new cluster
+    // This goes on until at last there is one cluster
+    // (n-1) + n = (2n-1)
+    parentVector.set_size(2*n-1);
+    parentVector.fill(SIZE_MAX);
+    sizeVector.set_size(2*n-1);
+    for (size_t i = 0; i < n; i++)
+      sizeVector(i) = 1;
+    for (size_t i = n; i < 2*n-1; i++)
+      sizeVector(i) = 0;
+    nextLabel = n;
+  }
+  void UnionOfTwoClusters(size_t m, size_t n)
+  {
+    // Each union will form a new cluster
+    // with size as sum of its children
+    sizeVector(nextLabel) = sizeVector(m) + sizeVector(n);
+    parentVector(m) = nextLabel;
+    parentVector(n) = nextLabel;
+    nextLabel++;
+  }
+  size_t Find(size_t n)
+  {
+    // finds the parent
+    size_t root = n;
+    while (parentVector(root) != SIZE_MAX)
+      root = parentVector(root);
+    while (parentVector(n) != root)
+    {
+      size_t temp = parentVector(n);
+      parentVector(n) = root;
+      n = temp;
+      if (n == SIZE_MAX)
+        break;
+    }
+      return root;
+  }
+  size_t GetSize(size_t i)
+  {
+    return sizeVector(i);
+  }
+};
+
+/**
+ * Converts the input (sorted MST) to a single linkage tree.
+ * Each row of the output contaisn 3 values
+ * (cluster1 cluster2 d size)
+ *  d = distance between cluster1 and cluster2
+ *  size = size of cluster formed by merging cluster1 and cluster2
+ *  cluster1 and cluster2 may contain more than one point
+ *  Initially points are labelled as 0,1,2,...,(n-1)
+ *  and are single cluster(contain only one point)
+ *  Then we go on merging 2 clusters at a time until we
+ *  reach a point when all points are in one cluster
+ * 
+ *  This is done with help of unionForSingleLinkageTree
+ *  For every edge in MST, a merge operation is performed
+ *  which merges two clusters.
+ */
+
+template<typename NeighborSearch,
+         typename MetricType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType >
+template<typename MatType>
+void HDBSCAN<NeighborSearch,
+             MetricType,
+             TreeType>::
+SingleLinkageTreeClustering(const MatType& inputMST,
+                          MatType& singleLinkageTree)
+{
+  singleLinkageTree.set_size(4, inputMST.n_cols);
+  size_t no = inputMST.n_cols + 1;
+  UnionForSingleLinkageClustering u(no);
+  size_t parentOfv1, parentOfv2;
+  double delta;
```

Comment:
Just a note, it can be a little hard for me (or another maintainer) to jump in and immediately understand what's going on here.  I figured it out after a little while, but it would be really helpful if you could go through and add some comments to some of the code.